.github/workflows/deploy.yml

```diff
@@ -0,0 +1,110 @@
+name: deploy
+
+on:
+  push:
+    branches: [main]
+    tags: ['v*.*.*']
+
+env:
+  AWS_ACCOUNT_ID: ${{ secrets.AWS_ACCOUNT_ID }}
+  AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+  AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+  AWS_DEFAULT_REGION: 'eu-central-1'
+  AWS_VPC_DEFAULT_SG_ID: ${{ secrets.AWS_VPC_DEFAULT_SG_ID }}
+  AWS_SUBNET_ID_1: ${{ secrets.AWS_SUBNET_ID_1 }}
+  AWS_SUBNET_ID_2: ${{ secrets.AWS_SUBNET_ID_2 }}
+  AWS_SUBNET_ID_3: ${{ secrets.AWS_SUBNET_ID_3 }}
+
+jobs:
+  init:
+    runs-on: ubuntu-latest
+    outputs:
+      environment: ${{ steps.setenv.outputs.environment}}
+
+    steps:
+      - name: Set environment
+        id: setenv
+        run: |
+          if [[ "${{github.base_ref}}" == "main" || "${{github.ref}}" == "refs/heads/main" ]]; then
+            echo "::set-output name=environment::testnet"
+          fi
+
+          if [[ "${{github.ref}}" == refs/tags/v* ]]; then
+            echo "::set-output name=environment::mainnet"
+          fi
+
+  deploy:
+    runs-on: ubuntu-latest
+    needs: init
+
+    environment: ${{ needs.init.outputs.environment }}
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.9
+      - name: Set up Node.js 14.x
+        uses: actions/setup-node@v1
+        with:
+          node-version: 14.x
+      - name: Cache node modules
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-node-modules
+        with:
+          # npm cache files are stored in `~/.npm` on Linux/macOS
+          path: /home/runner/work/hathor-explorer/hathor-explorer/node_modules
+          key: ${{ runner.os }}-build-${{ env.cache-name }}-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-build-${{ env.cache-name }}-
+            ${{ runner.os }}-build-
+            ${{ runner.os }}-
+      - name: Install Dependencies
+        run: |
+          pip -q --no-input install poetry
+          make install
+      - name: Linters
+        run: |
+          poetry run make check
+      - name: Deploy Lambdas Testnet
+        if: github.ref == 'refs/heads/main'
```

Comment:
Using those ifs is something that will probably come up during reviews, so I'll add my thoughts on this already.

There are some other options on how to do this with GIthub Actions, but I still thought this was the best one for now.

The others options would be:

### Create separate jobs for deploy on testnet and mainnet.
In this case, I would have to either repeat all the steps in each job, or try to share between them the steps that are equal.

I didn't thought repeating everything was good.

To share some steps, we would have to have another job with the shared steps, which are the dependencies installation, that produces an artifact in the end, which then we would download in the 'deployment' jobs.

The problem with this is that this artifact would have too much things inside it, like the node_modules and the python dependencies. Besides, `poetry` would need to be reinstalled in each job, or also be added in the artifact.

### Move every env var to the `init` job

This way, we could just have generic deploy steps here, and we would eliminate the `ifs`.

This is not a bad option, but it would probably require a few more things to work. Like possibly a third job, instead of doing it inside `init`, because some secrets depend on the job's `environment` being set, which is what the `init` job is doing.

Besides, the syntax would not be clean.

Reply:
I agree that this is the best solution for now. I was just thinking if there is a way to use the environment 'testnet' or 'mainnet' in the if, give that we already had this conditional check of the main branch or tag before.

Reply:
That's a good point. I think it's possible, will do this and test to confirm

Reply:
Confirmed, I have commited the changes